src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBioMartRetriever.java

```diff
@@ -61,27 +63,32 @@ public void downloadData() throws IOException, InterruptedException, BioMartQuer
-            // Iterate through each microarray type and retrieve Ensembl-Microarray identifier mappings.
-            // All mappings are stored in a single file, (eg: hsapiens_microarray_ids_and_go_terms);
-            for (String microarrayType : microarrayTypes) {
-                queryBioMartAndStoreData(speciesBioMartName, getBioMartXMLFilePath(), microarrayType, EnsemblBioMartUtil.MICROARRAY_IDS_AND_GO_TERMS_SUFFIX);
-            }
-
-            // Query GO and GO_SLIM identifier mapping data from BioMart and write it to a file (eg: hsapiens_microarray_ids_and_go_terms)
-            logger.info("Retrieving GO data");
-            for (String goQueryId : Arrays.asList(GO_ID_BIOMART_SEARCH_TERM, GO_SLIM_BIOMART_SEARCH_TERM)) {
-                queryBioMartAndStoreData(speciesBioMartName, getBioMartXMLFilePath(), goQueryId, EnsemblBioMartUtil.MICROARRAY_IDS_AND_GO_TERMS_SUFFIX);
+            biomartSearchTerms.addAll(queryBioMart(getMicroarrayTypesQuery(speciesBioMartName), MICROARRAY_TYPES));
+            // Iterate through each Biomart search term and retrieve Ensembl-otherIdentifier identifier mappings.
+            // All mappings are stored in a single file, (eg: hsapiens_microarray_go_ncbi_ids);
+            String otherIdentifiersFilename = this.destination + speciesBioMartName + EnsemblBioMartUtil.OTHER_IDENTIFIERS_SUFFIX;
+            if (!Files.exists(Paths.get(otherIdentifiersFilename))) {
+                for (String biomartSearchTerm : biomartSearchTerms) {
+                    logger.info("Retrieving " + biomartSearchTerm + " mappings from BioMart");
+                    queryBioMartAndStoreData(speciesBioMartName, getBioMartXMLFilePath(), biomartSearchTerm, otherIdentifiersFilename);
+                }
+            } else {
+                logFileExistsMessage(otherIdentifiersFilename);
             }
 
             // Query Ensembl-Uniprot (swissprot and trembl) identifier mapping data from BioMart and write it to a file (eg: hsapiens_uniprot).
-            logger.info("Retrieving UniProt data");
-            for (String uniprotQueryId : Arrays.asList(UNIPROT_SWISSPROT_BIOMART_SEARCH_TERM, UNIPROT_TREMBL_BIOMART_SEARCH_TERM)) {
-                queryBioMartAndStoreData(speciesBioMartName, getBioMartXMLFilePath(), uniprotQueryId, EnsemblBioMartUtil.UNIPROT_SUFFIX);
+            String uniprotFilename = this.destination + speciesBioMartName + EnsemblBioMartUtil.UNIPROT_SUFFIX;
+            if (!Files.exists(Paths.get(uniprotFilename))) {
+                logger.info("Retrieving UniProt mappings from BioMart");
+                for (String uniprotQueryId : Arrays.asList(UNIPROT_SWISSPROT_BIOMART_SEARCH_TERM, UNIPROT_TREMBL_BIOMART_SEARCH_TERM)) {
+                    queryBioMartAndStoreData(speciesBioMartName, getBioMartXMLFilePath(), uniprotQueryId, uniprotFilename);
+                }
+            } else {
+                logFileExistsMessage(uniprotFilename);
```

Comment:
The logic and structure of this block of code are almost identical to the BioMart search terms on lines 73-80.  Could you consolidate it in a more generic helper method that takes in the differences of the appropriate suffix (lines 73 and 84) and list of search terms (lines 75 and 87) as parameters?

Reply:
Yep, good catch